app/code/Magento/Indexer/Model/Processor.php

```diff
@@ -60,32 +65,83 @@ public function __construct(
      */
     public function reindexAllInvalid()
     {
-        $sharedIndexesComplete = [];
         foreach (array_keys($this->config->getIndexers()) as $indexerId) {
             /** @var Indexer $indexer */
             $indexer = $this->indexerFactory->create();
             $indexer->load($indexerId);
             $indexerConfig = $this->config->getIndexer($indexerId);
+            $sharedIndex = $indexerConfig['shared_index'];
+
             if ($indexer->isInvalid()) {
                 // Skip indexers having shared index that was already complete
                 $sharedIndex = $indexerConfig['shared_index'] ?? null;
-                if (!in_array($sharedIndex, $sharedIndexesComplete)) {
+                if (!in_array($sharedIndex, $this->sharedIndexesComplete)) {
                     $indexer->reindexAll();
-                } else {
-                    /** @var \Magento\Indexer\Model\Indexer\State $state */
-                    $state = $indexer->getState();
-                    $state->setStatus(StateInterface::STATUS_WORKING);
-                    $state->save();
-                    $state->setStatus(StateInterface::STATUS_VALID);
-                    $state->save();
-                }
-                if ($sharedIndex) {
-                    $sharedIndexesComplete[] = $sharedIndex;
+
+                    if ($sharedIndex) {
+                        $this->validateSharedIndex($sharedIndex);
+                    }
                 }
             }
         }
     }
 
+    /**
+     * Get indexer ids that have common shared index
+     *
+     * @param string $sharedIndex
+     * @return array
+     */
+    private function getIndexerIdsBySharedIndex(string $sharedIndex): array
+    {
+        $indexers = $this->config->getIndexers();
+
+        $result = [];
+        foreach ($indexers as $indexerConfig) {
+            if ($indexerConfig['shared_index'] == $sharedIndex) {
+                $result[] = $indexerConfig['indexer_id'];
+            }
+        }
+
+        return $result;
+    }
+
+    /**
+     * Validate indexers by shared index ID
+     *
+     * @param string $sharedIndex
+     * @return $this
+     */
+    private function validateSharedIndex(string $sharedIndex): self
+    {
+        if (empty($sharedIndex)) {
+            throw new \InvalidArgumentException(
```

Comment:
You are probably right, but this falls out of scope of this PR, it might fit better in the refactoring step which should happen when resolving https://github.com/magento/magento2/issues/29297

Sorry for being so strict, but I want to this bug to be fixed as soon as possible and not being held back by "code quality" requirements. That can happen later. The refactoring will probably not take that much time, but rewriting all the automated tests will take a significant amount of time and I don't have a lot of free time at the moment. Sorry!